spark3/src/main/java/org/apache/iceberg/spark/source/SparkMergeScan.java

```diff
@@ -0,0 +1,178 @@
+    this.snapshotId = Spark3Util.propertyAsLong(options, "snapshot-id", null);
+
+    Map<String, String> props = table.properties();
+
+    long tableSplitSize = PropertyUtil.propertyAsLong(props, SPLIT_SIZE, SPLIT_SIZE_DEFAULT);
+    this.splitSize = Spark3Util.propertyAsLong(options, "split-size", tableSplitSize);
+
+    int tableSplitLookback = PropertyUtil.propertyAsInt(props, SPLIT_LOOKBACK, SPLIT_LOOKBACK_DEFAULT);
+    this.splitLookback = Spark3Util.propertyAsInt(options, "lookback", tableSplitLookback);
+
+    long tableOpenFileCost = PropertyUtil.propertyAsLong(props, SPLIT_OPEN_FILE_COST, SPLIT_OPEN_FILE_COST_DEFAULT);
+    this.splitOpenFileCost = Spark3Util.propertyAsLong(options, "file-open-cost", tableOpenFileCost);
+
+    if (snapshotId == null) {
+      // init files with an empty list if the snapshot id is not set to avoid picking any concurrent changes
+      files = Collections.emptyList();
+    }
+  }
+
+  Long snapshotId() {
+    return snapshotId;
+  }
+
+  @Override
+  public Statistics estimateStatistics() {
+    if (snapshotId == null) {
+      return new Stats(0L, 0L);
+    }
+    return super.estimateStatistics();
+  }
+
+  // should be accessible to the write
+  List<FileScanTask> files() {
+    if (files == null) {
+      TableScan scan = table
+          .newScan()
+          .caseSensitive(caseSensitive())
+          .useSnapshot(snapshotId)
+          .project(expectedSchema);
+
+      for (Expression filter : filterExpressions()) {
+        scan = scan.filter(filter);
+      }
+
+      if (ignoreResiduals) {
+        scan.ignoreResiduals();
+      }
+
+      try (CloseableIterable<FileScanTask> filesIterable = scan.planFiles()) {
+        this.files = Lists.newArrayList(filesIterable);
+      } catch (IOException e) {
+        throw new RuntimeIOException(e, "Failed to close table scan: %s", scan);
+      }
+    }
+
+    return files;
+  }
+
+  @Override
+  protected List<CombinedScanTask> tasks() {
+    if (tasks == null) {
+      CloseableIterable<FileScanTask> splitFiles = TableScanUtil.splitFiles(
+          CloseableIterable.withNoopClose(files()),
+          splitSize);
+      CloseableIterable<CombinedScanTask> scanTasks = TableScanUtil.planTasks(
+          splitFiles, splitSize,
+          splitLookback, splitOpenFileCost);
+      tasks = Lists.newArrayList(scanTasks);
+    }
+
+    return tasks;
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    SparkMergeScan that = (SparkMergeScan) o;
+    return table().name().equals(that.table().name()) &&
+        readSchema().equals(that.readSchema()) && // compare Spark schemas to ignore field ids
+        filterExpressions().toString().equals(that.filterExpressions().toString()) &&
+        ignoreResiduals == that.ignoreResiduals &&
+        Objects.equals(snapshotId, that.snapshotId);
```

Comment:
We need to remember to add the filtered file list here after that is supported, right?

Reply:
Yeah, we would need to add it here once we have `SupportsFileFilter` interface.

Reply:
Let me add a TODO item here so that we don't forget.